### PR TITLE
Enable Cortex-R82 Userspace

### PR DESCRIPTION
**Pull request by povergoing**
Enables the Userspace feature for the [Cortex-R82](https://developer.arm.com/ip-products/processors/cortex-r/cortex-r82) processor. 
The Cortex-R82 supports the 64bit MPU, so the Userspace is based on MPU.

This PR includes:
1. Refine the current code, and unify the MMU and MPU interfaces.
2. Add the Userspace support based on **AArch64 MPU**
3. Fix the test cases

Currently, the only board that supports **AArch64 MPU** is [fvp_baser_aemv8r](https://docs.zephyrproject.org/latest/boards/arm64/fvp_baser_aemv8r/doc/index.html).

I have passed all 318 test cases include the userspace relevant test cases. 
```
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testcase list...
INFO    - 1464 test scenarios (1404 configurations) selected, 1086 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:  318/ 318  100%  skipped: 1165, failed:    0
INFO    - 239 of 239 test configurations passed (100.00%), 0 failed, 1165 skipped with 0 warnings in 379.65 seconds
INFO    - In total 1896 test cases were executed, 17633 skipped on 1 out of total 394 platforms (0.25%)
INFO    - 192 test configurations executed on platforms, 47 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing xunit report /home/jaxhan01/src/zephyrproject_rebase/twister-out/twister.xml...
INFO    - Writing xunit report /home/jaxhan01/src/zephyrproject_rebase/twister-out/twister_report.xml...
INFO    - Run completed
```
Here I uploaded the test report [twister_report.zip](https://github.com/zephyrproject-rtos/zephyr/files/7159321/twister_report.zip)

### CODE REVIEW

File: arch/arm64/core/cortex_r/arm_mpu.c
Comment (carlocaione) [inline arch/arm64/core/cortex_r/arm_mpu.c:233]: nit: can you move this at the beginning of the function?

File: arch/arm64/core/cortex_r/arm_mpu.c
Comment (carlocaione) [inline arch/arm64/core/cortex_r/arm_mpu.c:345]: nit: pick a consistent type for the `for` cycles. You used `int` here but `uint8_t` previously. Also decide if you want to declare the variable inside the `for` or outside. In general my suggestion is always to use `size_t`.

**Comment by carlocaione** [comment]
I'm not familiar with the MPU so besides a few nits LGTM but I can only give a shallow review and trust the test framework.

File: include/arch/arm64/cortex_r/arm_mpu.h
Comment (dcpleung) [inline include/arch/arm64/cortex_r/arm_mpu.h:154]: Is there a particular reason you want the RAM region to be visible to everyone? The design of userspace is for `.data` section to be only visible to supervisor threads. User threads must be explicitly allowed to access data via memory domains. This is a security consideration to avoid user threads to read arbitrary data, especially those that must not be exposed.

File: soc/arm64/arm/fvp_aemv8r/arm_mpu_regions.c
Comment (dcpleung) [inline soc/arm64/arm/fvp_aemv8r/arm_mpu_regions.c:40]: `_app_smem_start` is only for symbols in memory partitions, which should have different permissions than `.data`.

File: scripts/gen_app_partitions.py
Comment (dcpleung) [inline scripts/gen_app_partitions.py:93]: Forcing an alignment here may create a memory hole for an empty section, depending on linker configuration. If this is required for ARMv8R, please make this own for ARMv8R.

File: arch/arm64/core/cortex_r/arm_mpu.c
Comment (povergoing) [reply arch/arm64/core/cortex_r/arm_mpu.c:233]: Sure, I will

File: arch/arm64/core/cortex_r/arm_mpu.c
Comment (povergoing) [reply arch/arm64/core/cortex_r/arm_mpu.c:252]: yes

File: arch/arm64/core/cortex_r/arm_mpu.c
Comment (povergoing) [reply arch/arm64/core/cortex_r/arm_mpu.c:345]: Ok, I'll change it to size_t

File: include/arch/arm64/cortex_r/arm_mpu.h
Comment (povergoing) [reply include/arch/arm64/cortex_r/arm_mpu.h:154]: Sorry for the unclear name, the `REGION_RAM_TEXT_ATTR` is for the `.text` section. I am not sure, but this config exactly follows the MMU https://github.com/zephyrproject-rtos/zephyr/blob/79952ad774caf380fca4d24cd12f397b9c337d29/arch/arm64/core/mmu.c#L650 
And so does the following RO section (P_RO_U_RO) https://github.com/zephyrproject-rtos/zephyr/blob/79952ad774caf380fca4d24cd12f397b9c337d29/arch/arm64/core/mmu.c#L656

File: soc/arm64/arm/fvp_aemv8r/arm_mpu_regions.c
Comment (povergoing) [reply soc/arm64/arm/fvp_aemv8r/arm_mpu_regions.c:40]: This is only for the init stage, and both `_app_smem_start` and `.data` are inited as `P_RW_U_NA`. These sections act as background/default regions (actually they will be marked as `dyn_region` after MPU being inited), and then they will be reconfigured during the thread init and context-switching dynamically.

File: scripts/gen_app_partitions.py
Comment (povergoing) [reply scripts/gen_app_partitions.py:93]: I agree, but this is located in python script. I'll try to find a better way to solve it.

File: arch/arm64/core/cortex_r/arm_mpu.c
Comment (alexanderwachter) [inline arch/arm64/core/cortex_r/arm_mpu.c:17]: Hint: the log level can go to this macro as a parameter

File: arch/arm64/core/cortex_r/arm_mpu.c
Comment (alexanderwachter) [inline arch/arm64/core/cortex_r/arm_mpu.c:27]: Nit: Why are these macros prefixed with an underscore?

File: soc/arm64/arm/fvp_aemv8r/arm_mpu_regions.c
Comment (dcpleung) [reply soc/arm64/arm/fvp_aemv8r/arm_mpu_regions.c:40]: Just for clarification, will `.data` have its own permissions or will it be grouped together with shared member?

File: scripts/gen_app_partitions.py
Comment (dcpleung) [reply scripts/gen_app_partitions.py:93]: Since this script is generating a linker script, `#ifdef` would work.

File: arch/arm64/core/cortex_r/arm_mpu.c
Comment (povergoing) [reply arch/arm64/core/cortex_r/arm_mpu.c:17]: Thanks, I will fix this.

File: arch/arm64/core/cortex_r/arm_mpu.c
Comment (povergoing) [reply arch/arm64/core/cortex_r/arm_mpu.c:27]: I thought it is only for the inside of this file. Anyway, I'll remove the underscore

File: scripts/gen_app_partitions.py
Comment (povergoing) [reply scripts/gen_app_partitions.py:93]: No problem.

File: soc/arm64/arm/fvp_aemv8r/arm_mpu_regions.c
Comment (povergoing) [reply soc/arm64/arm/fvp_aemv8r/arm_mpu_regions.c:40]: The `.data` and `_app_smem_start` will be reconfiged as their own separated permissions. I remember there are some test cases that will test this permission.

**Review by carlocaione — APPROVED** [review]

File: soc/arm64/arm/fvp_aemv8r/arm_mpu_regions.c
Comment (povergoing) [reply soc/arm64/arm/fvp_aemv8r/arm_mpu_regions.c:40]: @dcpleung Would you like to have a look again?:)

**Review by dcpleung — APPROVED** [review]

**Comment by povergoing** [comment]
Hi @carlocaione, may I ask what else I can do before merging?

**Comment by carlocaione** [comment]
> Hi @carlocaione, may I ask what else I can do before merging?

Nothing. Just waiting for the merge window to be open again.

**Pull request merged**
